langchain/src/llms/googlepalm.ts

```diff
@@ -0,0 +1,208 @@
+  get lc_secrets(): { [key: string]: string } | undefined {
+    return {
+      apiKey: "GOOGLEPALM_API_KEY",
+    };
+  }
+
+  model = "models/text-bison-001";
+
+  temperature?: number = undefined; // default value chosen based on model
+
+  maxOutputTokens?: number = undefined; // defaults to 64
+
+  topP?: number = undefined; // default value chosen based on model
+
+  topK?: number = undefined; // default value chosen based on model
+
+  stopSequences: string[] = [];
+
+  safetySettings?:
+    | protos.google.ai.generativelanguage.v1beta2.ISafetySetting[] = undefined; // default safety setting for that category
+
+  apiKey?: string = undefined;
+
+  private client: TextServiceClient;
+
+  constructor(fields?: GooglePalmTextInput) {
+    super(fields ?? {});
+
+    this.model = fields?.model ?? this.model;
+    if (this.model && !this.model.startsWith("models/")) {
+      throw new Error(
+        "`model` value must follow the pattern - `models/{model}`"
+      );
+    }
+
+    this.temperature = fields?.temperature ?? this.temperature;
+    if (this.temperature && (this.temperature < 0 || this.temperature > 1)) {
+      throw new Error("`temperature` must be in the range of [0.0,1.0]");
+    }
+
+    this.maxOutputTokens = fields?.maxOutputTokens ?? this.maxOutputTokens;
+    if (this.maxOutputTokens && this.maxOutputTokens < 0) {
+      throw new Error("`maxOutputTokens` must be a positive integer");
+    }
+
+    this.topP = fields?.topP ?? this.topP;
+    if (this.topP && this.topP < 0) {
+      throw new Error("`topP` must be a positive integer");
+    }
+
+    this.topK = fields?.topK ?? this.topK;
+    if (this.topK && this.topK < 0) {
+      throw new Error("`topK` must be a positive integer");
+    }
+
+    this.stopSequences = fields?.stopSequences ?? this.stopSequences;
+
+    this.safetySettings = fields?.safetySettings ?? this.safetySettings;
+    if (this.safetySettings && this.safetySettings.length > 0) {
+      const safetySettingsSet = new Set(
+        this.safetySettings.map((s) => s.category)
+      );
+      if (safetySettingsSet.size !== this.safetySettings.length) {
+        throw new Error(
+          "The categories in `safetySettings` array must be unique"
+        );
+      }
+    }
+
+    this.apiKey =
+      fields?.apiKey ?? getEnvironmentVariable("GOOGLEPALM_API_KEY");
+    if (!this.apiKey) {
+      throw new Error(
+        "Please set an API key for Google Palm 2 in the environment variable GOOGLEPALM_API_KEY or in the `apiKey` field of the GooglePalm constructor"
+      );
+    }
+
+    this.client = new TextServiceClient({
+      authClient: new GoogleAuth().fromAPIKey(this.apiKey),
+    });
+  }
+
+  _llmType(): string {
+    return "googlepalm";
+  }
+
+  async _call(
+    prompt: string,
+    options: this["ParsedCallOptions"]
+  ): Promise<string> {
+    const res = await this.caller.callWithOptions(
+      { signal: options.signal },
```

Comment:
Is it possible to pass this into the client directly? You *can* do this but it's a last resort - it's hacky and doesn't actually cancel the in-flight request.

Reply:
I tried this out, unfortunately, I can't find a reliable way with Google's clients. They use [`google-gax`](https://github.com/googleapis/gax-nodejs) -> [`grpc-js`](https://github.com/grpc/grpc-node) lib which has a similar issue open here - https://github.com/grpc/grpc-node/issues/1603. I'm not an expert on this topic :(. Need help here. The `client` has a `TextServiceClient::close()` but that completely terminates the `client` connections and becomes no longer usable.

Reply:
No worries then.